File: frontend/types/training.ts
/**
 * Training types for RPE tracking and session load management
 */

export type RpeUpsertPayload = {
  player_id: string;
  session_id: string;
  rpe: number; // 0..10
};

export type RpeUpsertResponse = {
  player_id: string;
  session_id: string;
  rpe: number;
  duration_min: number;
  session_load: number;
  updated_at: string;
};

export type WeeklyLoadPoint = {
  week_start: string;      // ISO date (YYYY-MM-DD)
  weekly_load: number;     // sum session_load in that week
};

export type WeeklyLoadResponse = {
  player_id: string;
  weeks: WeeklyLoadPoint[];
  total_current_week: number;
};

export type TrainingSession = {
  id: string;
  session_date: string;
  session_type: string;
  duration_min: number;
  focus?: string;
  description?: string;
};

File: backend/app/schemas/training.py
"""Training schemas for RPE tracking and session load calculation."""

from datetime import date, datetime
from decimal import Decimal
from typing import List
from uuid import UUID

from pydantic import BaseModel, Field, field_validator


class RpeUpsertPayload(BaseModel):
    """Payload for upserting RPE data for a player session."""
    player_id: UUID
    session_id: UUID
    rpe: Decimal = Field(..., ge=0, le=10, decimal_places=1)

    @field_validator('rpe')
    @classmethod
    def validate_rpe(cls, v: Decimal) -> Decimal:
        """Ensure RPE is between 0 and 10 with max 1 decimal place."""
        if v < 0 or v > 10:
            raise ValueError('RPE must be between 0 and 10')
        return v


class RpeUpsertResponse(BaseModel):
    """Response after upserting RPE data."""
    player_id: UUID
    session_id: UUID
    rpe: Decimal
    duration_min: int
    session_load: Decimal
    updated_at: datetime

    class Config:
        from_attributes = True


class WeeklyLoadPoint(BaseModel):
    """Single week data point for weekly load chart."""
    week_start: date
    weekly_load: Decimal

    class Config:
        from_attributes = True


class WeeklyLoadResponse(BaseModel):
    """Response containing weekly load aggregations for a player."""
    player_id: UUID
    weeks: List[WeeklyLoadPoint]
    total_current_week: Decimal

    class Config:
        from_attributes = True
